src/server/pkg/storage/fileset/merge.go

```diff
@@ -381,26 +496,21 @@ func shard(mr *MergeReader, shardThreshold int64, f ShardFunc) error {
 	var size int64
 	pathRange := &index.PathRange{}
 	if err := mr.Iterate(func(fmr *FileMergeReader) error {
-		if pathRange.Lower == "" {
-			pathRange.Lower = fmr.Index().Path
-		}
-		size += fmr.Index().SizeBytes
 		// A shard is created when we have encountered more than shardThreshold content bytes.
 		if size >= shardThreshold {
 			pathRange.Upper = fmr.Index().Path
 			if err := f(pathRange); err != nil {
 				return err
 			}
 			size = 0
-			pathRange = &index.PathRange{}
+			pathRange = &index.PathRange{
```

Comment:
This is just a cleaner way of doing this, right? Not meaningful for this change?

Reply:
It is cleaner, but it also was necessary for handling deletions correctly. The merge reader does not return deleted files, but we need to apply the deletions during compaction. An easy case to see where the prior algorithm does not work is if the last shard should just be deletions. In this case, the shard would not get created because the callback would not be executed for that shard (since the merge reader won't return anything for deleted files). This makes sure that the full path range of the input filesets is accounted for.